README
vprerex (visual prerex) is a GUI front-end and viewer for the prerex
interactive editor of prerequisite-chart descriptions in the prerex
format. The viewer may be used just as a minimalist viewer of arbitrary
PDF files but supports editing of charts generated using prerex.sty by
allowing coordinates of course boxes, arrows, and background points
to be conveyed back to the prerex editor command-line using the X11
clipboard.

To install:

  + You must have first installed the qt-5 and poppler libraries,
    including the poppler-qt5 library. These may be available packaged
    for your platform or the sources can be obtained from:

http://qt.nokia.com/downloads/downloads#qt-lib
http://poppler.freedesktop.org/

  + Edit vprerex.pro so INCLUDEPATH and LIBS point to the 
    relevant directories.

  + qmake vprerex.pro (using the qt-4 qmake)

  + make

  + install the vprerex binary in a suitable directory such as
    $HOME/bin or /usr/local/bin.

If vprerex is applied to a LaTeX file, it calls xterm and the prerex
editor.



ChangeLog
Version 6.5.1 2017-11-26

  - modified for qt5 and poppler-qt5

Version 6.4.4 2014-10-21

  - added a file-opening dialog to specify a chart file
  - allow for creation of a new "null" prerex file

Version 6.4.3 2014-10-17

 - check that the xterm has started
 - set xterm font size to 10pt
 
Version 6.4.2 2012-05-30

 - allow for Windows (rather than X11) clipboard function

Version 6.4.1 2012-05-25

 - restore FileSystemWatch path after reloading pdf (needed for Windows)

Version 6.4.0 2012-04-13

 - references to undeclared sleep() replaced by references to QThread::sleep()  
   (debian bug 667331).

Version 6.3.2 2012-03-30

 - bump up version number to match prerex.sty and prerex

Version 6.3.1 2011-12-22

  If a file opening dialogue is used and a tex file is opened, the Working Directory 
  for the prerex process is set to the directory of the tex file. 

Version 6.3 2011-11-22

  Right-clicking on a course box or arrow *appends* coordinates to the
  clipboard (to simplify multiple deletes/shifts).

Version 6.2 2011-09-08

  no longer embedding the prerex editing window because of continuing focus issues

Version 6.1.1 2011-08-20

 - using grabKeyboard to avoid keyboard-focus issues
 - bump up version number to match prerex.sty and prerex

Version 6.0.1 2011-07-27

 - modified FocusPolicy

Version 6.0  2010-06-23

 - files re-organized and distributed as a gzipped tarball

Version 5.7  2010-04-06

  - corrected coordinate calculations to use the image width/height rather than
    the widget width/height (and allow for image offsets)

Version 5.6  2010-03-26
Version 5.5  2010-03-16
 
  - re-named vprerex (visual prerex)
  - opens prerex source files by embedding a prerex command-line window
  - added Quit button to controls
  - removed docking features from the controls widget 
  - removed resizing from the statusbar

  This is the first distributed version of vprerex; the version numbers are
  starting at 5.5 to be compatible with the current version of prerex.

  Derived from pdfviewer by

    - excising all UI widgets except the Page spinner and the Scale selector

    - adding URI tooltips and cursor-shape changes for links

    - assigning box, edge, and background-point coordinates to the X clipboard 
      when the user clicks on a box, edge mid-point, or background point, resp. 
      (for a prerex chart with the coordinate grid enabled)
   
    - adding a file watcher to re-load the PDF when the file changes

    - posting taskbar messages to identify the program version and the loaded file,
      and inform the user that the PDF has been re-loaded




main.cpp
/****************************************************************************
** Adapted from PDFviewer, originally published in the QT Quarterly:
** http://doc.trolltech.com/qq/QtQuarterly27.pdf
** Copyright (C) 2008 Nokia Corporation and/or its subsidiary(-ies).
** Contact: Qt Software Information (qt-info@nokia.com)
**
** Modified for use with prerex:
** http://www.ctan.org/tex-archive/graphics/prerex
** Copyright (C) 2009-14 R. D. Tennent 
** School of Computing, Queen's University, rdt@cs.queensu.ca 

** This program is free software; you can redistribute it and/or modify it
** under the terms of the GNU General Public License as published by the
** Free Software Foundation; either version 2 of the License, or (at your
** option) any later version.
** 
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
** Public License for more details.
** 
** You should have received a copy of the GNU General Public License along
** with this program; if not, write to the Free Software Foundation, Inc.,
** 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
**
****************************************************************************/

/*
** vprerex (visual prerex) is a GUI front-end and viewer for the prerex
** interactive editor of prerequisite-chart descriptions in the prerex
** format. The viewer may be used just as a minimalist viewer of
** arbitrary PDF files but supports editing of charts generated using
** prerex.sty by allowing coordinates of course boxes, arrows, and
** background points to be conveyed back to the prerex editor command-line
** using the clipboard.
*/

#include <QApplication>
#include "window.h"
#include <iostream>

using namespace std;


int main(int argc, char *argv[])
{

    QApplication app(argc, argv);
    Window window;
    QStringList Argv = app.arguments();

    printf ("This is vprerex, version %s.\n", version);
    puts ("Copyright (c) 2008 Nokia Corporation and/or its subsidiary(-ies).");
    puts ("Qt Software Information (qt-info@nokia.com)");
    puts ("Copyright (c) 2009-17  R. D. Tennent");
    puts ("School of Computing, Queen's University, rdt@cs.queensu.ca");
    puts
      ("This program comes with NO WARRANTY. You can redistribute vprerex and/or");
    puts
      ("modify it under the terms of the GNU General Public License version 2.");
    puts ("");

    window.show();
    if ( argc > 1 )
    {
      QString fileName;
      QString chartFileName;
      fileName.append(Argv.at(1));
      if (!fileName.endsWith(".tex")  &&  !fileName.endsWith(".pdf")) 
        fileName.append(".tex");
      if ( argc > 2 )
      {
        chartFileName.append(Argv.at(2));
        if (!chartFileName.endsWith(".tex")) 
          chartFileName.append(".tex");
      }
      window.setFileNames(fileName, chartFileName);
    }
    else 
      window.open();

    return app.exec();
}


window.h

/****************************************************************************
** Adapted from PDFviewer, originally published in the QT Quarterly:
** http://doc.trolltech.com/qq/QtQuarterly27.pdf
** Copyright (C) 2008 Nokia Corporation and/or its subsidiary(-ies).
** Contact: Qt Software Information (qt-info@nokia.com)
**
** Modified for use with prerex:
** http://www.ctan.org/tex-archive/graphics/prerex
** Copyright (C) 2009-14 R. D. Tennent 
** School of Computing, Queen's University, rdt@cs.queensu.ca 

** This program is free software; you can redistribute it and/or modify it
** under the terms of the GNU General Public License as published by the
** Free Software Foundation; either version 2 of the License, or (at your
** option) any later version.
** 
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
** Public License for more details.
** 
** You should have received a copy of the GNU General Public License along
** with this program; if not, write to the Free Software Foundation, Inc.,
** 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
**
****************************************************************************/

#ifndef WINDOW_H
#define WINDOW_H

# include <QMainWindow>
# include "documentwidget.h"
# include <QFileSystemWatcher>
# include "ui_window.h"
# include <QProcess>
# include <QMessageBox>
# include <QFileDialog>


const char version[] = "6.5.1";

class QTextBrowser;

class Window : public QMainWindow, public Ui_MainWindow
{
    Q_OBJECT

public:
    Window(QWidget *parent = 0);
    QFileSystemWatcher *fsw;
    QStatusBar *sb;
    void setFileNames(QString BaseFileName, QString ChartFileName);
    void loadDocument(void);
    void open();

private slots:
    void scaleDocument(int index);
    void reloadMessage();
    void pathMessage();
    void quit();

private:
    DocumentWidget *documentWidget;
    QString fileName;
    QFileInfo fileInfo;
    QString chartFileName;
    QFileInfo chartFileInfo;
    QString pdfFilePath;
    QVector<qreal> scaleFactors;
    QLabel *versionWidget;
    QProcess *process;
};

#endif


documentwidget.h

/****************************************************************************
** Adapted from PDFviewer, originally published in the QT Quarterly:
** http://doc.trolltech.com/qq/QtQuarterly27.pdf
** Copyright (C) 2008 Nokia Corporation and/or its subsidiary(-ies).
** Contact: Qt Software Information (qt-info@nokia.com)
**
** Modified for use with prerex:
** http://www.ctan.org/tex-archive/graphics/prerex
** Copyright (C) 2009-12 R. D. Tennent 
** School of Computing, Queen's University, rdt@cs.queensu.ca 

** This program is free software; you can redistribute it and/or modify it
** under the terms of the GNU General Public License as published by the
** Free Software Foundation; either version 2 of the License, or (at your
** option) any later version.
** 
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
** Public License for more details.
** 
** You should have received a copy of the GNU General Public License along
** with this program; if not, write to the Free Software Foundation, Inc.,
** 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
**
****************************************************************************/

#ifndef DOCUMENTWIDGET_H
#define DOCUMENTWIDGET_H

#include <QRectF>
#include <QLabel>
#include <poppler-qt5.h>
#include <qclipboard.h>
#include <QFileInfo>
#include <QToolTip>
#include <QMessageBox>

class DocumentWidget : public QLabel
{
    Q_OBJECT

public:
    DocumentWidget(QWidget *parent = 0);
    ~DocumentWidget();
    Poppler::Document *document();
    QMatrix matrix() const;
    qreal scale() const;
    bool setDocument(const QString &filePath);

public slots:
    void reloadDocument1();
    void reloadDocument2();
    void setPage(int page = -1);
    void setScale(qreal scale);

protected:
    void mousePressEvent(QMouseEvent *event);
    void mouseMoveEvent(QMouseEvent *event);
    void mouseReleaseEvent(QMouseEvent *event);

signals:
    void pageChanged(int currentPage);
    void documentReloaded();

private:
    void showPage(int page = -1);

    Poppler::Document *doc;
    int currentPage;
    qreal scaleFactor;
    QList<Poppler::Link * > pageLinks;
    bool mousePressed;
    QClipboard *cb;
    QString currentPath;
    QTimer *timer ;
    QFileInfo *fileInfo;
    int imageWidth;
    int imageHeight;
    int imageXoffset;
    int imageYoffset;
};

#endif


documentwidget.cpp

/****************************************************************************
** Adapted from PDFviewer, originally published in the QT Quarterly:
** http://doc.trolltech.com/qq/QtQuarterly27.pdf
** Copyright (C) 2008 Nokia Corporation and/or its subsidiary(-ies).
** Contact: Qt Software Information (qt-info@nokia.com)
**
** Modified for use with prerex:
** http://www.ctan.org/tex-archive/graphics/prerex
** Copyright (C) 2009-14 R. D. Tennent 
** School of Computing, Queen's University, rdt@cs.queensu.ca 

** This program is free software; you can redistribute it and/or modify it
** under the terms of the GNU General Public License as published by the
** Free Software Foundation; either version 2 of the License, or (at your
** option) any later version.
** 
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
** Public License for more details.
** 
** You should have received a copy of the GNU General Public License along
** with this program; if not, write to the Free Software Foundation, Inc.,
** 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
**
****************************************************************************/

#include <QtGui>
#include <poppler-qt5.h>
#include "documentwidget.h"

DocumentWidget::DocumentWidget(QWidget *parent)
    : QLabel(parent)
{
    currentPage = -1;
    doc = NULL;
    scaleFactor = 1.1;
    setCursor(Qt::ArrowCursor);
    setAlignment(Qt::AlignTop | Qt::AlignCenter);
    mousePressed = false;
    setMouseTracking(true);
    cb = QGuiApplication::clipboard();
    timer = new QTimer(this);
    connect(timer, SIGNAL(timeout()), this, SLOT(reloadDocument2()));
}

DocumentWidget::~DocumentWidget()
{
    delete doc;
}

Poppler::Document *DocumentWidget::document()
{
    return doc;
}

void DocumentWidget::mousePressEvent(QMouseEvent *event)
{
    Qt::MouseButton qtm = event->button();
    if ( !(qtm == Qt::LeftButton || qtm == Qt::RightButton) ) return;
    int x1 = 0, y1 = 0, x2 = 0, y2 = 0;     // pixel coordinates of anchors 
    int xc1 = 0, yc1 = 0, xc2 = 0, yc2 = 0; // chart coordinates of anchors
    int n_anchors = 0;   // number of anchors found 
    imageXoffset = 0; imageYoffset = 0;     // re-compute in case of window re-sizing
    if (imageWidth < width())
       imageXoffset = (width() - imageWidth) / 2;
    if (imageHeight < height())  // possible after re-scaling despite AlignTop
       imageYoffset = (height() - imageHeight) / 2;
    int xmouse = event->pos().x() - imageXoffset;  // mouse coordinates in pixels
    int ymouse = event->pos().y() - imageYoffset;
    QPointF mousePos  // normalized position in [0..1] coordinates
     = QPointF((qreal)xmouse / (qreal)imageWidth, (qreal)ymouse / (qreal)imageHeight);

    
    // iterate through the Links:
    for (int i = 0; i < pageLinks.size(); ++i)
    {
        Poppler::LinkBrowse * link = static_cast<Poppler::LinkBrowse * >(pageLinks[i]);
        if (link == NULL) continue;  // not a Browse Link
        QString URI = link->url();
        QRectF linkRect = pageLinks[i]->linkArea();
        if ( linkRect.contains(mousePos) &&  URI.startsWith(QString("coord:")) )
            {   // clicked in a coord: x,y box or coord: x0,y0,x1,y1 bullet
                QString cbString;
                if ( qtm == Qt::RightButton )
                {
                   if (cb->supportsSelection())
                     cbString.append(cb->text(QClipboard::Selection));
                   else
                     cbString.append(cb->text(QClipboard::Clipboard));
                   cbString.append(" ");
                }
                URI.remove("coord: ");
                cbString.append(URI);
                if (cb->supportsSelection())
                  cb->setText(cbString, QClipboard::Selection);
                else
                  cb->setText(cbString, QClipboard::Clipboard);
                setCursor(Qt::CrossCursor);
                mousePressed = true;
                return;
            }
        if ( URI.startsWith(QString("anchor:")) ) 
        {   
            QPointF anchorCenter  //  in [0..1] coordinates 
            = (pageLinks[i]->linkArea()).center();
            URI.remove(0,7);      // remove "anchor:"
            int comma = URI.indexOf(',');
            QString xs = URI.left(comma);        // x-coordinate as a string
            QString ys = URI.remove(0, comma+1); // y-coordinate as a string
            if (n_anchors == 0)
            {
               xc1 = xs.toInt();
               yc1 = ys.toInt();
               x1 = (int)(imageWidth * anchorCenter.x());
               y1 = (int)(imageHeight * anchorCenter.y());
               n_anchors++;
            }
            else if (n_anchors == 1)
            {
               xc2 = xs.toInt();
               yc2 = ys.toInt();
               x2 = (int)(imageWidth * anchorCenter.x());
               y2 = (int)(imageHeight * anchorCenter.y());
               n_anchors++;
            }
        }
    }
    if (n_anchors == 0) return;  // coordinate grid off?
    if (n_anchors == 1)          
    {
      QMessageBox::warning (this, "vprerex", "Only found one anchor");
      return;
    }
    //  clicked on background grid; must be LeftButton click
    if ( qtm == Qt::RightButton ) return;
    //  determine chart coordinates of the click point:
    int  xc = (int)( 0.25L + (xc1 + xc2)/2.0L - ((x1 + x2)/2.0L - xmouse) * (qreal)(xc2 - xc1) / (x2 - x1) );
    int  yc = (int)( 0.75L + (yc1 + yc2)/2.0L - ((y1 + y2)/2.0L - ymouse) * (qreal)(yc2 - yc1) / (y2 - y1) );
    if (cb->supportsSelection())
      cb->setText(QString("%1,%2").arg(xc).arg(yc), QClipboard::Selection);
    else
      cb->setText(QString("%1,%2").arg(xc).arg(yc), QClipboard::Clipboard);
    mousePressed = true;
    setCursor(Qt::CrossCursor);
}

void DocumentWidget::mouseMoveEvent(QMouseEvent *event)
{

    imageXoffset = 0; imageYoffset = 0;  // re-compute in case of window re-sizing
    if (imageWidth < width())
       imageXoffset = (width() - imageWidth) / 2;
    if (imageHeight < height())  // possible after re-scaling despite AlignTop
       imageYoffset = (height() - imageHeight) / 2;
    int xmouse = event->pos().x() - imageXoffset;  // mouse coordinates in pixels
    int ymouse = event->pos().y() - imageYoffset;
    QPointF mousePos  // normalized click position in [0..1] coordinates
    = QPointF((qreal)xmouse / (qreal)imageWidth, (qreal)ymouse / (qreal)imageHeight);
    
    // iterate through the Links:
    for (int i = 0; i < pageLinks.size(); ++i)
    {
        Poppler::LinkBrowse * link = static_cast<Poppler::LinkBrowse * >(pageLinks[i]);
        if (link == NULL) continue;  // not a Browse Link
        QString URI = link->url();
        QRectF linkRect = pageLinks[i]->linkArea();
        if ( linkRect.contains(mousePos) ) 
        {
          if ( mousePressed ) return;
          setCursor(Qt::PointingHandCursor);
          setToolTip(URI);
          return;
        }
    }
    if (mousePressed) return;
    setCursor(Qt::ArrowCursor);
}

void DocumentWidget::mouseReleaseEvent(QMouseEvent *)
{
    mousePressed = false;
    setCursor(Qt::ArrowCursor);
}

qreal DocumentWidget::scale() const
{
    return scaleFactor;
}

void DocumentWidget::showPage(int page)
{
    if (page != -1 && page != currentPage + 1) {
        currentPage = page - 1;
        emit pageChanged(page);
    }
    QImage image 
    = doc->page(currentPage)->renderToImage(scaleFactor * physicalDpiX(), scaleFactor * physicalDpiY());
    setPixmap(QPixmap::fromImage(image));
    imageWidth = image.width();
    imageHeight = image.height();
    pageLinks = doc -> page(currentPage) -> links();
    setCursor(Qt::ArrowCursor);
}

bool DocumentWidget::setDocument(const QString &filePath)
{
    doc = Poppler::Document::load(filePath);
    if (doc == NULL) {
    // temporarily ill-formed pdf?  wait for another reload signal
      return( false );
    }
    currentPath = filePath;
    fileInfo = new QFileInfo(currentPath);
    doc->setRenderHint(Poppler::Document::Antialiasing);
    doc->setRenderHint(Poppler::Document::TextAntialiasing);
    currentPage = -1;
    setPage(1);
    return( true );
}

void DocumentWidget::reloadDocument1(void)
{ 
  fileInfo->refresh();
  if (fileInfo->size() == 0) return; // pdflatex clears file.pdf 
  timer->setSingleShot(true);
  timer->start(2000);  // wait for the file to stabilize
  // When the timer timesout, it calls reloadDocument2
}

void DocumentWidget::reloadDocument2(void)
{
    Poppler::Document *savedoc;
    fileInfo->refresh();
    if (fileInfo->size() == 0) return; // pdflatex clears file.pdf 
    savedoc = doc;
    doc = Poppler::Document::load(currentPath);
    if (doc == NULL) {
    // temporarily ill-formed pdf?  wait for another reload signal
        doc = savedoc;
        return;
    }
    delete savedoc;
    doc->setRenderHint(Poppler::Document::Antialiasing);
    doc->setRenderHint(Poppler::Document::TextAntialiasing);
    showPage(currentPage+1);
    documentReloaded();
}

void DocumentWidget::setPage(int page)
{
    if (page != currentPage + 1) {
        showPage(page);
    }
}

void DocumentWidget::setScale(qreal scale)
{
    if (scaleFactor != scale) {
        scaleFactor = scale;
        showPage(-1);
    }
}


window.cpp

/****************************************************************************
** Adapted from PDFviewer, originally published in the QT Quarterly:
** http://doc.trolltech.com/qq/QtQuarterly27.pdf
** Copyright (C) 2008 Nokia Corporation and/or its subsidiary(-ies).
** Contact: Qt Software Information (qt-info@nokia.com)
**
** Modified for use with prerex:
** http://www.ctan.org/tex-archive/graphics/prerex
** Copyright (C) 2009-14 R. D. Tennent 
** School of Computing, Queen's University, rdt@cs.queensu.ca 

** This program is free software; you can redistribute it and/or modify it
** under the terms of the GNU General Public License as published by the
** Free Software Foundation; either version 2 of the License, or (at your
** option) any later version.
** 
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
** Public License for more details.
** 
** You should have received a copy of the GNU General Public License along
** with this program; if not, write to the Free Software Foundation, Inc.,
** 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
**
****************************************************************************/
#include <QtGui>
#include "window.h"


# include <QString>
# include <QStringList>
# include <QThread>
# include <QInputDialog>

#include <qthread.h>

QString x_terminal_emulator = "xterm" ;

class mythread : public QThread
{
public:
        static void sleep(unsigned long secs) {
                QThread::sleep(secs);
        }
};


Window::Window(QWidget *parent)
    : QMainWindow(parent)
{
    setupUi(this);

    scaleFactors << 1. << 1.1 << 1.25 << 1.75 << 2.5 << 4.;

    documentWidget = new DocumentWidget();
    scrollAreaWidget->setWidget(documentWidget);
    fsw = new QFileSystemWatcher();
    sb = statusBar();
    versionWidget = new QLabel(QString("This is vprerex, v.%1        ").arg(version), this);
    sb->addPermanentWidget(versionWidget);

    connect(documentControlsAction, SIGNAL(toggled(bool)),
            controlsDockWidget, SLOT(setVisible(bool)));

    connect(pageSpinBox, SIGNAL(valueChanged(int)),
            documentWidget, SLOT(setPage(int)));
    connect(documentWidget, SIGNAL(pageChanged(int)),
            pageSpinBox, SLOT(setValue(int)));
    connect(scaleComboBox, SIGNAL(currentIndexChanged(int)),
            this, SLOT(scaleDocument(int)));


    connect(fsw, SIGNAL(fileChanged(QString)), 
            documentWidget, SLOT(reloadDocument1()));
    connect(documentWidget, SIGNAL(documentReloaded()),
            this, SLOT(reloadMessage()));
    connect(quitButton, SIGNAL(clicked()), this, SLOT(quit()));

}

void Window::setFileNames(QString FileName, QString ChartFileName )
{
  fileName.append(FileName);
  fileInfo.setFile(FileName);
  chartFileName.append(ChartFileName);
  chartFileInfo.setFile(ChartFileName);
  loadDocument();
}

void Window::open()
{
   QString fullFileName;
   QString fullChartFileName;
   fullFileName.append(QDir::cleanPath(QFileDialog::getOpenFileName(this, "Select a prerex file to edit or a pdf file to view (or Cancel).", QString(), "Prerex source files (*.tex);;Portable Document Format files (*.pdf)")));
   if (fullFileName.length() == 0)   
   { // file selection cancelled 
     bool ok;
     QFile file("tmp.tex");
     while (1)
     {
       fullFileName.append(QDir::cleanPath(QFileDialog::getExistingDirectory(this, "Select a working directory.")));
       if ( fullFileName.length() == 0 ) { QMessageBox::critical(this, "vprerex", "Directory selection failed."); close(); exit(1);
       }
       fullFileName.append("/");
       fullFileName.append(QInputDialog::getText(this, "Creating a new prerex file", "Input file name:", QLineEdit::Normal, "mychart.tex" , &ok)); 
       if (!ok) 
       {    
         QMessageBox::critical(this, "vprerex", "File name input failed."); close(); exit(1);
       }
       if (!fullFileName.endsWith(".tex")) 
          fullFileName.append(".tex");
       file.setFileName(fullFileName);
       if (!file.exists()) break;
       if ( QMessageBox::question(this, "vprerex", "File exists; Ok to overwrite?",
              QMessageBox::Ok | QMessageBox::No, QMessageBox::Ok) == QMessageBox::Ok) break; 
       fullFileName.clear();
     }  
     if (!file.open(QIODevice::WriteOnly | QIODevice::Text))
     {    
       QMessageBox::critical(this, "vprerex", "File opening failed."); close(); exit(1);
     }
     QTextStream out(&file);
     out << "\\documentclass{article}\n" << "\\usepackage{geometry}\n" << "\\geometry{noheadfoot, margin=0.5in}\n" << "\\usepackage{prerex}\n" << "\\begin{document}\n" << "\\thispagestyle{empty}\n" << "\\begin{chart}\n" << "\\end{chart}\n" << "\\end{document}\n";
     file.close();
     fileInfo.setFile(fullFileName);
     fileName.append(fileInfo.baseName());
     fileName.append(".");
     fileName.append(fileInfo.suffix());
   }
   else
   {
     fileInfo.setFile(fullFileName);
     fileName.append(fileInfo.baseName());
     fileName.append(".");
     fileName.append(fileInfo.suffix());
     fullChartFileName.append(QDir::cleanPath(QFileDialog::getOpenFileName(this, "Select a chart file (or Cancel).", fileInfo.absolutePath(), "Prerex chart files (*.tex)")));
     if (fullChartFileName.length() > 0)
     {
       chartFileInfo.setFile(fullChartFileName);
       chartFileName.append(chartFileInfo.baseName());
       chartFileName.append(".");
       chartFileName.append(chartFileInfo.suffix());
     }
   }
   loadDocument();
}

void Window::loadDocument(void)
{
   pdfFilePath.append(fileInfo.dir().path());
   pdfFilePath.append("/");
   pdfFilePath.append(fileInfo.baseName());
   pdfFilePath.append(".pdf");
   if ( fileName.endsWith(".tex") )
   {
     QStringList arguments;
     arguments   // for xterm
      << "-geometry" << "100x60" 
      << "-fa" << "Monospace"
      << "-fs" << "10"
      << "-e" << "prerex"; 
     arguments.append(fileName);
     if ( chartFileName.length() > 0 )
       arguments.append(chartFileName);
     process = new QProcess(centralwidget);
     process->setWorkingDirectory(fileInfo.dir().path());
     process->start(x_terminal_emulator, arguments);  
     if (!process->waitForStarted()) 
     { 
       QMessageBox::critical(this, "vprerex", "x-terminal call failed.");
       close();
       exit(1);
     }
   }

   QFileInfo pdfFileInfo;
   pdfFileInfo.setFile(pdfFilePath);
   while ( pdfFileInfo.size() == 0 ) 
   { // wait for the PDF file 
        mythread::sleep (1);
        pdfFileInfo.refresh();
   }
   mythread::sleep (1);
   sb->showMessage(pdfFilePath);
   fsw->addPath(pdfFilePath);
   while ( !(documentWidget->setDocument(pdfFilePath)) ) 
   { // if at first you don't succeed, try, try again 
     mythread::sleep (1);
   }    
   scaleComboBox->setEnabled(true);
   pageSpinBox->setEnabled(true);
   pageSpinBox->setMinimum(1);
   pageSpinBox->setMaximum(documentWidget->document()->numPages());
   pageSpinBox->setValue(1);
}

void Window::reloadMessage(void)
{
  sb->showMessage(pdfFilePath + " reloaded.", 2000);
  QTimer::singleShot(2000, this, SLOT(pathMessage()));
  
  // On Windows, pdflatex removes the existing pdf file initially and
  // this removes the file path from the file-system watcher; the
  // following restores the path to the file-watch list if necessary:
  if (!fsw->files().contains(pdfFilePath)) fsw->addPath(pdfFilePath); }

void Window::pathMessage(void)
{
  sb->showMessage(pdfFilePath);
}

void Window::scaleDocument(int index)
{
  documentWidget->setScale(scaleFactors[index]);
}

void Window::quit(void)
{
  close();
}
